# Kickstarter and SQL `10 points`

Since I claimed that the point of Lede was to be able to learn anything using Google, I should probably test that, too, yeah?

Working with **SQL databases** is something we didn't learn! It comes in a lot of formats - PostreSQL, MySQL, Sqlite - but they are all generally the same thing.

SQL is great for when you're reading in big big big datasets: instead of waiting an hour for `.read_csv` to crash on a giant file that you're going to filter anyway, instead you just ask SQL to give you a subset of the data to work on.

**Topics**

* Learning new things with the power of the internet
* SQL

### The data

**Kickstarter** is a website people use to raise money for projects. It used to be really popular, but it's since died down a bit. We're going to work with [a dataset about Kickstarter in SQLite format](https://dataverse.harvard.edu/dataset.xhtml?persistentId=doi:10.7910/DVN/EOYBXM).

You'll want to download the database by going to **Access Dataset** and then **Download ZIP**.

It's a big file, and it's compressed! We'll be using the `kickstarter.db` file that is inside the zip file: when you open up the zip, I extracted it using [The Unarchiver](https://theunarchiver.com/) by telling it to open up the `.z01` file (it's so big it's spread across both the `.zip` and the `.z01`). 

## Open the dataset

Just like an Excel file can have different sheets, a SQL database has different tables.

In [31]:
import pandas as pd

pd.set_option("display.max_columns", None)
pd.set_option('display.float_format', '{:,.2f}'.format)

SQLite + pandas doesn't use the "normal" way to ask for tables, so I'll give that line to you.

In [2]:
import sqlite3

# Open a connection
conn = sqlite3.connect("kickstarter.db")

# Make a query, save it into a dataframe
df = pd.read_sql_query('SELECT name FROM sqlite_master WHERE type = "table";', conn)

# Close the connection
conn.close()

# See the results in the dataframe
df

name
0         urls_to_scrape
1              all_files
2               category
3                creator
4          funding_trend
5                   item
6             livestream
7               location
8                project
9                 reward
10                update
11              comments
12  creator_external_url
13           reward_item
14           creator_bio

## Simple selects `3 points`

### Select all of the data for the first 4 kickstarter projects

They live in the `project` table. This will help you understand how SQL queries work.

In [21]:
conn = sqlite3.connect("kickstarter.db")
df1 = pd.read_sql_query("SELECT * from project LIMIT 4", conn)
conn.close()
df1

id       state                                        url_project  \
0  22807353    canceled  https://www.kickstarter.com/projects/mikewood/...   
1  22810934  successful  https://www.kickstarter.com/projects/cavinboun...   
2  22821161      failed  https://www.kickstarter.com/projects/684178251...   
3  22823613      failed  https://www.kickstarter.com/projects/123754320...   

                                   url_project_short  \
0                              http://kck.st/1UNDF6C   
1                              http://kck.st/17xiZX3   
2  https://www.kickstarter.com/projects/684178251...   
3                              http://kck.st/2kL47iy   

                                                name country  creator_id  \
0  ViperSharp, The Best Precision Knife Sharpener...      US  1750652465   
1                              Lights, Coma, Action!      US  1157451606   
2        Sentio Golf Putters: Feel IS the difference      US   684178251   
3                                           Brainade      US  1237543205   

   location_id  category_id  created_at    deadline  updated_at  \
0     23418188           28  1453170931  1462402800  1472168774   
1      2357536           48  1380594589  1383413613  1488327585   
2      2351810           28  1433725049  1446429290  1463739059   
3      2497409          342  1486073091  1489778686  1489778687   

   state_changed_at    successful_at  launched_at   goal   pledged currency  \
0        1461010001              NaN   1458918033  10000 31,131.00      USD   
1        1383413613 1,383,413,613.00   1380821613  15000 18,167.88      USD   
2        1446429290              NaN   1442537690  50000  9,707.00      USD   
3        1489778687              NaN   1487190286    750      1.00      USD   

  currency_symbol  usd_pledged  static_usd_rate  backers_count  \
0               $    31,131.00                1            255   
1               $    18,167.88                1            284   
2               $     9,707.00                1             41   
3               $         1.00                1              1   

   comments_count  updates_count spotlight staff_pick  \
0              53             22     False      False   
1               4             40      True      False   
2               0              4     False      False   
3               0              0     False      False   

                                               blurb currency_trailing_code  \
0  The ViperSharp offers infinite angles for shar...                   True   
1  A book to tell the story of who I was (LIGHTS)...                   True   
2  Choose the feel YOU want with our patented flo...                   True   
3  We intend to create an online platform for ind...                   True   

  disable_communication                                          photo_url  \
0                 False  https://ksr-ugc.imgix.net/assets/012/357/781/2...   
1                 False  https://ksr-ugc.imgix.net/assets/011/584/530/2...   
2                 False  https://ksr-ugc.imgix.net/assets/012/161/151/b...   
3                 False  https://ksr-ugc.imgix.net/assets/015/387/055/f...   

  profile_background_color  profile_background_image_opacity  \
0                     None                              0.80   
1                     None                              0.80   
2                     None                              0.80   
3                     None                              0.80   

                                       profile_blurb  profile_id  \
0                                               None     2328882   
1  A book to tell the story of how I sustained a ...      716409   
2                                               None     1945377   
3                                               None     2862898   

  profile_link_background_color           profile_link_text  \
0                          None                        None   
1                          None 

### Select the top 10 projects with the highest amount pledged

I only want to see the name and the amount pledged. **The number one result should be the Pebble watch** - if your result is something Japanese, try querying all of the columns and figure out what went wrong.

In [30]:
conn = sqlite3.connect("kickstarter.db")
df2 = pd.read_sql_query('select name, usd_pledged from project order by usd_pledged desc limit 10', conn)
conn.close()
df2

name   usd_pledged
0   Pebble Time - Awesome Smartwatch, No Compromises 20,338,986.27
1  COOLEST COOLER: 21st Century Cooler that's Act... 13,285,226.36
2             Pebble 2, Time 2 + All-New Pebble Core 12,779,843.49
3                         Kingdom Death: Monster 1.5 12,393,139.69
4       Pebble: E-Paper Watch for iPhone and Android 10,266,845.74
5  The World's Best TRAVEL JACKET with 15 Feature...  9,192,055.66
6                                  Exploding Kittens  8,782,571.99
7             OUYA: A New Kind of Video Game Console  8,596,474.58
8  THE 7th CONTINENT – What Goes Up, Must Come Down.  7,072,757.00
9             The Everyday Backpack, Tote, and Sling  6,565,782.50

## Filtering your queries `3 points`

### Find me – yes, me, Soma! – in the project creators table

In [58]:
conn = sqlite3.connect("kickstarter.db")
df3 = pd.read_sql_query("select * from creator WHERE name like 'Jonathan Soma'", conn)
conn.close()
df3

id           name  slug  \
0  546599331  Jonathan Soma  soma   

                                        avatar_thumb  \
0  https://ksr-ugc.imgix.net/assets/005/801/114/d...   

                                        avatar_small  \
0  https://ksr-ugc.imgix.net/assets/005/801/114/d...   

                              urls_web_user  \
0  https://www.kickstarter.com/profile/soma   

                                       avatar_medium is_registered  \
0  https://ksr-ugc.imgix.net/assets/005/801/114/d...          None   

                                       urls_api_user chosen_currency  
0  https://api.kickstarter.com/v1/users/546599331...            None

### Find all of my projects

In [57]:
conn = sqlite3.connect("kickstarter.db")
df4 = pd.read_sql_query("select * from project WHERE creator_id like '546599331'", conn)
conn.close()
df4

id       state                                        url_project  \
0   320516934  successful  https://www.kickstarter.com/projects/soma/broo...   
1  1052891075  successful  https://www.kickstarter.com/projects/soma/cook...   
2  1606179513  successful  https://www.kickstarter.com/projects/soma/ice-...   

      url_project_short                                             name  \
0  http://kck.st/d9d9Sx                  Brooklyn Brainery Needs a Home!   
1  http://kck.st/aZ8hta  Cooking Class in a Box [Exotic Cuisine Edition]   
2  http://kck.st/cgpgNy                                   Ice Cream Club   

  country  creator_id  location_id  category_id  created_at    deadline  \
0      US   546599331     12589335            1  1270872639  1278129540   
1      US   546599331     12589335           10  1276719823  1278991380   
2      US   546599331      2459115           10  1271981825  1274055420   

   updated_at  state_changed_at  successful_at  launched_at  goal  pledged  \
0  1397755136        1278129615     1278129615   1271182430  9500 9,629.95   
1  1397756801        1278991807     1278991807   1276792614   200 1,820.44   
2  1397755379        1274056206     1274056206   1272034766   220   485.98   

  currency currency_symbol  usd_pledged  static_usd_rate  backers_count  \
0      USD               $     9,629.95                1            204   
1      USD               $     1,820.44                1             53   
2      USD               $       485.98                1             27   

   comments_count  updates_count spotlight staff_pick  \
0               9              9      True       True   
1               8              7      True       True   
2               7              5      True       True   

                                               blurb currency_trailing_code  \
0  We host awesomely cheap, collaborative classes...                   True   
1  I'll build you a cooking class on an exotic cu...                   True   
2  When it's 100 degrees outside you'll wish you ...                   True   

  disable_communication                                          photo_url  \
0                 False  https://ksr-ugc.imgix.net/assets/011/261/121/e...   
1                 False  https://ksr-ugc.imgix.net/assets/011/263/591/4...   
2                 False  https://ksr-ugc.imgix.net/assets/011/261/487/c...   

  profile_background_color  profile_background_image_opacity profile_blurb  \
0                     None                              0.80          None   
1                     None                              0.80          None   
2                     None                              0.80          None   

   profile_id profile_link_background_color profile_link_text  \
0        4894                          None              None   
1        7857                          None              None   
2        5368                          None              None   

  profile_link_text_color profile_link_url profile_name  profile_project_id  \
0                    None             None         None                4894   
1                    None             None         None                7857   
2                    None             None         None                5368   

  profile_should_show_feature_image_section profile_show_feature_image  \
0                                      True                      False   
1                                      True                      False   
2                                      True                      False   

  profile_state  profile_state_changed_at profile_text_color  \
0      inactive                1425915800               None   
1      inactive                1425915800               None   
2      inactive                1425915800               None   

                                            slug  \
0                 brooklyn-brainery-needs-a-home   
1  cooking-class-in-a-box-exotic-cuisine-edition   
2         

## Calculations `3 points`

### Find how much money has been raised on Kickstarter by **successful** projects.

In [56]:
conn = sqlite3.connect("kickstarter.db")
df5 = pd.read_sql_query("select SUM(usd_pledged) from project WHERE state like 'successful'", conn)
conn.close()
df5

SUM(usd_pledged)
0  3,643,252,201.72

### Find how much money has been raised on Kickstarter by "successful" projects in **the United States**.

The `state` can be successful, failed, live, etc.

In [55]:
conn = sqlite3.connect("kickstarter.db")
df6 = pd.read_sql_query("select SUM(usd_pledged) from project WHERE state like 'successful' AND country like 'US'", conn)
conn.close()
df6

SUM(usd_pledged)
0  2,961,700,473.92

## Aggregating `3 points`

## How many projects were successful/unsuccessful/etc?

**Use your SQL query to calculate the answer.** Your result should look something like this:

| |...|...|
|---|---|---|
|0|canceled|36659|
|1|failed|211783|
|2|live|3163|
|3|purged|202|
|4|successful|155130|
|5|suspended|1700|

In [62]:
conn = sqlite3.connect("kickstarter.db")
df7 = pd.read_sql_query("select state, COUNT(state) from project Group by state", conn)
conn.close()
df7

state  COUNT(state)
0    canceled         36659
1      failed        211783
2        live          3163
3      purged           202
4  successful        155130
5   suspended          1700

### How much money has been pledged to campaigns in each country?

In [89]:
conn = sqlite3.connect("kickstarter.db")
df8 = pd.read_sql_query("select country, SUM(usd_pledged) from project Group by country", conn)
conn.close()
df8

country  SUM(usd_pledged)
0       AT     14,451,653.89
1       AU     54,965,224.87
2       BE      5,963,505.71
3       CA    112,432,907.01
4       CH     19,435,417.64
5       DE     54,424,475.45
6       DK     11,033,924.53
7       ES     24,281,129.46
8       FR     48,536,026.19
9       GB    298,732,943.23
10      HK     22,263,452.47
11      IE      5,678,333.70
12      IT     25,627,131.64
13      JP      4,676,375.59
14      LU        546,185.66
15      MX      4,091,229.60
16      NL     36,543,906.82
17      NO      3,897,921.52
18      NZ     10,581,166.47
19      SE     20,266,478.73
20      SG      9,507,867.91
21      US  3,270,883,676.02

## Conversion to CSV `2 points`

### Save the result of your last query - the one about amount raised - to a CSV

In [90]:
df8.to_csv('kickstarter_campaigns_by_country.csv')

### Visualize the results using DataWrapper, highlighting the country of your choice

Link in the cell below.

https://www.datawrapper.de/_/n2orc/

## Joins

### Show all the columns in the `category` table

In [79]:
conn = sqlite3.connect("kickstarter.db")
df9 = pd.read_sql_query("pragma table_info('category');", conn, index_col='cid')
conn.close()
df9

name          type  notnull dflt_value  pk
cid                                                         
0                   id           int        1       None   1
1                 name  varchar(190)        0       None   0
2            parent_id           int        0       None   0
3             position           int        0       None   0
4                color           int        0       None   0
5                 slug  varchar(190)        0       None   0
6    urls_web_discover          text        0       None   0

### Show all the columns in the `project` table

In [80]:
conn = sqlite3.connect("kickstarter.db")
df10 = pd.read_sql_query("pragma table_info('project');", conn, index_col='cid')
conn.close()
df10

name             type  notnull  \
cid                                                                        
0                                           id              int        1   
1                                        state      varchar(20)        0   
2                                  url_project     varchar(190)        0   
3                            url_project_short     varchar(190)        0   
4                                         name     varchar(190)        0   
5                                      country      varchar(20)        0   
6                                   creator_id              int        0   
7                                  location_id              int        0   
8                                  category_id              int        0   
9                                   created_at              int        0   
10                                    deadline              int        0   
11                                  updated_at              int        0   
12                            state_changed_at              int        0   
13                               successful_at              int        0   
14                                 launched_at              int        0   
15                                        goal   decimal(20, 4)        0   
16                                     pledged   decimal(12, 4)        0   
17                                    currency      varchar(20)        0   
18                             currency_symbol      varchar(20)        0   
19                                 usd_pledged  decimal(20, 10)        0   
20                             static_usd_rate  decimal(20, 10)        0   
21                               backers_count              int        0   
22                              comments_count              int        0   
23                               updates_count              int        0   
24                                   spotlight      varchar(20)        0   
25                                  staff_pick      varchar(20)        0   
26                                       blurb             text        0   
27                      currency_trailing_code      varchar(20)        0   
28                       disable_communication      varchar(20)        0   
29                                   photo_url             text        0   
30                    profile_background_color     varchar(190)        0   
31            profile_background_image_opacity   decimal(12, 4)        0   
32                               profile_blurb             text        0   
33                                  profile_id              int        0   
34               profile_link_background_color      varchar(20)        0   
35                           profile_link_text             text        0   
36                     profile_link_text_color      varchar(20)        0   
37                            profile_link_url             text        0   
38                                profile_name     varchar(190)        0   
39                          profile_project_id              int        0   
40   profile_should_show_feature_image_section      varchar(20)        0   
41                  profile_show_feature_image      varchar(20)        0   
42                               profile_state      varchar(20)        0   
43                    profile_state_changed_at              int        0   
44                          profile_text_color      varchar(20)        0   
45                                        slug     varchar(190)        0   
46                                 url_rewards     varchar(190)        0   
47                                 url_updates     varchar(190)        0   
48                                    video_id              int        0   
49                              video_url_high     varchar(190)        0   
50                              video_url_webm     varchar(190)        0   
51            

### Using SQL, request the first 10 projects and merge in the category of each project `3 points`

In [88]:
conn = sqlite3.connect("kickstarter.db")
df11 = pd.read_sql_query("SELECT project.name as Project, category.name as Category from project LEFT JOIN category ON category.id = project.category_id LIMIT 10", conn)
conn.close()
df11

Project          Category
0  ViperSharp, The Best Precision Knife Sharpener...    Product Design
1                              Lights, Coma, Action!        Nonfiction
2        Sentio Golf Putters: Feel IS the difference    Product Design
3                                           Brainade               Web
4                                  Horizon Anthology       Anthologies
5                 The Alphabet Story Children's Book  Children's Books
6  Torchlight Lullaby: A Fantasy Adventure Graphi...    Graphic Novels
7            The Gears LP, from scratch... With You.        Indie Rock
8              Polynesian Adventure - the board game    Tabletop Games
9                                 Masquerading Masks               Art

### Building on your last query, calculate the number of projects in each category `2 points`

In [99]:
conn = sqlite3.connect("kickstarter.db")
df12 = pd.read_sql_query("SELECT category.name as Category, count(category.name) as Number_of_Projects from project LEFT JOIN category ON category.id = project.category_id group by category.name"  , conn)
conn.close()
df12

Category  Number_of_Projects
0    3D Printing                 781
1       Academic                1062
2    Accessories                4542
3         Action                 837
4        Animals                 281
..           ...                 ...
154  Woodworking                1311
155    Workshops                 176
156  World Music                2265
157  Young Adult                 939
158        Zines                 478

[159 rows x 2 columns]

### Using pandas, sort the result to get the top 20 project categories `1 point`

In [100]:
df12.sort_values("Number_of_Projects", ascending = False).head(20)

Category  Number_of_Projects
113    Product Design               24270
90              Music               17055
136    Tabletop Games               16702
39        Documentary               15961
129            Shorts               12572
148       Video Games               12492
58               Food               11054
55       Film & Video               10523
54            Fiction                9510
10                Art                8725
95         Nonfiction                8623
52            Fashion                8303
7             Apparel                8299
19   Children's Books                7477
8                Apps                7198
141           Theater                7147
125              Rock                6971
138        Technology                6034
153         Webseries                5855
72         Indie Rock                5831